pkg/gofr/datasource/opentsdb/response.go

```diff
@@ -0,0 +1,416 @@
+	span := c.addTrace(ctx, "isValidOperateMethod")
+
+	status := statusSuccess
+
+	var message string
+
+	defer sendOperationStats(c.logger, time.Now(), "isValidOperateMethod", &status, &message, span)
+
+	method = strings.TrimSpace(strings.ToUpper(method))
+	if method == "" {
+		return false
+	}
+
+	validMethods := []string{http.MethodPost, http.MethodPut, http.MethodDelete, http.MethodPut}
+	for _, validMethod := range validMethods {
+		if method == validMethod {
+			return true
+		}
+	}
+
+	return false
+}
+
+func customParserHelper(ctx context.Context, resp genericResponse, operation string, logger Logger, tracer trace.Tracer,
+	unmarshalFunc func([]byte) error) func([]byte) error {
+	span := resp.addTrace(ctx, tracer, operation)
+
+	status := statusFailed
+
+	var message string
+
+	defer sendOperationStats(logger, time.Now(), operation, &status, &message, span)
+
+	return func(result []byte) error {
+		err := unmarshalFunc(result)
+		if err != nil {
+			message = fmt.Sprintf("unmarshal %s error: %s", operation, err)
+			logger.Errorf(message)
+
+			return err
+		}
+
+		status = statusSuccess
+		message = fmt.Sprintf("%s custom parsing was successful.", operation)
+
+		return nil
+	}
+}
+
+func queryParserHelper(ctx context.Context, logger Logger, tracer trace.Tracer, obj genericResponse,
+	methodName string) func(respCnt []byte) error {
+	return customParserHelper(ctx, obj, methodName, logger, tracer, func(resp []byte) error {
+		originRespStr := string(resp)
+
+		var respStr string
+
+		if strings.HasPrefix(string(resp), "[") && strings.HasSuffix(string(resp), "]") {
+			respStr = fmt.Sprintf(`{"queryRespCnts":%s}`, originRespStr)
+		} else {
+			respStr = originRespStr
+		}
+
+		return json.Unmarshal([]byte(respStr), obj)
+	})
+}
+
+func (c *Client) operateAnnotation(ctx context.Context, queryAnnotation, resp any, method, operation string) error {
+	span := c.addTrace(ctx, operation)
+
+	status := statusFailed
+
+	var message string
+
+	defer sendOperationStats(c.logger, time.Now(), operation, &status, &message, span)
+
+	annotation, ok := queryAnnotation.(*Annotation)
+	if !ok {
+		return errors.New("invalid annotation type. Must be *Annotation")
+	}
+
+	annResp, ok := resp.(*AnnotationResponse)
+	if !ok {
+		return errors.New("invalid response type. Must be *AnnotationResponse")
+	}
+
+	if !c.isValidOperateMethod(ctx, method) {
+		message = fmt.Sprintf("invalid annotation operation method: %s", method)
+		return errors.New(message)
+	}
+
+	annoEndpoint := fmt.Sprintf("%s%s", c.endpoint, annotationPath)
+
+	resultBytes, err := json.Marshal(annotation)
+	if err != nil {
+		message = fmt.Sprintf("marshal annotation response error: %s", err)
+		return errors.New(message)
+	}
+
+	if err = c.sendRequest(ctx, method, annoEndpoint, string(resultBytes), annResp); err != nil {
+		message = fmt.Sprintf("%s: error while processing %s annotation request to url %q: %s", operation, method, annoEndpoint, err.Error())
+		return err
```

Comment:
I would recommend the following pattern everywhere


```suggestion
		err = fmt.Errorf("%s error while processing %s annotation request to url %q: %w", operation, method, annoEndpoint, err.Error)
		message = err.Error()
		return err
```

@Umang01-hash what do you think?

Error should be wrapped with information especially in the error returned and not only the log

Reply:
GoFr has different log levels .....message is actually a debug message only appearing at debug level, which I feel must give information about the action in which the operation is being performed, So, I don't feel the context of the error should be the same as the debug level. It will diminish the value of debug level, if user gets all the info through errors....

But I am open to suggestions....